### PR TITLE
A fix in the core typechecker

### PR DESCRIPTION
**Pull request by aseemr**
Adding a missing check for expected effect in the case when the expected type is not set.

The fix failed some examples, where the core client wanted a total effect on non-informative types, but core computed a ghost effect (and the cases passed due to the missing check).

To fix those cases, the PR also adds eager promotion so that it computes a total effect in such cases.

The PR also contains a small optimization in inserting hide/reveal coercions. If the expression is `reveal e` (resp. `hide e`), and the context expects an erased (resp. non-erased) type, then instead of making the coercion return `hide (reveal e)` (resp. `reveal (hide e)`, we just return `e`.

The last optimization resulted in two failures, one each in `Steel.ST.Array` and `Steel.ST.HigherArray`. Since the coerced expression is a little different, two `rewrite` calls in the libraries (one in each file) had to be tweaked. In general, smt attribute saves us, but these cases were where the coercion was happening in the first argument of `pts_to` (see https://github.com/FStarLang/steel/pull/63).

### CODE REVIEW

File: examples/tactics/eci19/ConstructiveLogic.fst [hunk 1]
```diff
@@ -14,13 +14,14 @@ commonly, however, the programmer needs to give the SMT solver some
 help in the form of new facts that it can use. We've seen examples of
 this when we call lemmas within a function body in order to make it
 verify. For instance, compare the following two examples (the first of
-which fails).
+which in the comments is known to be unstable).
 *)
 
-[@expect_failure]
+(*
```
Comment (aseemr) [inline examples/tactics/eci19/ConstructiveLogic.fst:20]: This is interesting. The example that was showing failing smt proofs, now works.

Comment (mtzguido) [reply examples/tactics/eci19/ConstructiveLogic.fst:20]: Weird! I noticed the same thing in a completely separate branch, so maybe some dependency got updated? In fact even master does not pass (local) ci for me right now.

**Review by mtzguido — APPROVED** [review]

File: examples/tactics/eci19/ConstructiveLogic.fst (hunk 1 above)
Comment (mtzguido) [reply examples/tactics/eci19/ConstructiveLogic.fst:20]: There seems to be a difference between running with --query_stats and without. CI will pass using that flag (this definition will fail as expected). You might need to revert this change to get a green. I'll look to see what causes this.

Comment (mtzguido) [reply examples/tactics/eci19/ConstructiveLogic.fst:20]: Solved in #3010

**Comment by aseemr** [comment]
I have an everest green, merging.

**Pull request merged**